### PR TITLE
Fix part of #1966: Fix InputInteractionViewTestActivityTest for Robolectric

### LINKED ISSUE #1966: Fix InputInteractionViewTestActivityTest for Robolectric
Currently, the InputInteractionViewTestActivityTest are passing on Espresso but they should also pass on Robolectric. Fix these below mentioned files to make sure that the tests pass on Robolectric.

You can run robolectric tests by following these steps: https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#running-roboelectric-tests

Fixing test cases will need to introduce TextApplicationComponent. This can be seen [here](https://github.com/oppia/oppia-android/pull/1680/commits/4729d551b86282ce617e1d53f2762416758785a1).

### PR DESCRIPTION
**Pull request by prayutsu**
## Explanation

Fixes part of #1966 :
Failing test cases on robolectric are fixed

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Roboletric
Screenshot from 2020-11-18 21-59-20

Espresso - 
Screenshot from 2020-11-18 22-18-57

### CODE REVIEW

**Comment by prayutsu** [comment]
@MohamedMedhat1998 PTAL

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: This test case is failing

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt:872]: This test case is failing

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt:756]: This test case is failing

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [reply app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: This is the error - 
```
androidx.test.espresso.base.DefaultFailureHandler$AssertionFailedWithCauseError: 'with string from resource id: <2131755343>' doesn't match the selected view.
Expected: with string from resource id: <2131755343>[ratio_error_invalid_format] value:
**** Please enter a valid ratio (e.g. 1:2 or 1:2:3). ****
     Got: "AppCompatTextView{id=2131296812, res-name=ratio_input_error, visibility=VISIBLE, width=888, height=96, has-focus=false, has-focusable=false, has-window-focus=true, is-clickable=false, is-enabled=true, is-focused=false, is-focusable=false, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@7be72b1a, tag=null, root-is-layout-requested=false, has-input-connection=false, x=60.0, y=532.0, 
***  text=Please write a ratio that consists of digits separated by colons (e.g. 1:2 or 1:2:3). ***
, input-type=0, ime-target=false, has-links=false}"
```

I have highlighted the expected and actual texts with ****

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [reply app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt:872]: Error - 
```
androidx.test.espresso.base.DefaultFailureHandler$AssertionFailedWithCauseError: 'with text: is "1:2:3:4"' doesn't match the selected view.

Expected: with text: is "1:2:3:4"

     Got: "RatioInputInteractionView{id=2131296953, res-name=test_ratio_input_interaction_view, visibility=VISIBLE, width=912, height=144, has-focus=false, has-focusable=true, has-window-focus=true, is-clickable=true, is-enabled=true, is-focused=false, is-focusable=true, is-layout-requested=false, is-selected=false, layout-params=android.widget.LinearLayout$LayoutParams@3ef666e3, tag=null, root-is-layout-requested=false, has-input-connection=true, editor-info=[inputType=0x1 imeOptions=0xc000006 privateImeOptions=null actionLabel=null actionId=0 initialSelStart=16 initialSelEnd=16 initialCapsMode=0x0 hintText=Enter a ratio in the form x:y. label=null packageName=null fieldId=0 fieldName=null extras=null hintLocales=null contentMimeTypes=null ], x=24.0, y=364.0, *** text=1   : 2 : 3 : 4 ***,
 hint=Enter a ratio in the form x:y., input-type=1, ime-target=false, has-links=false}"

```

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [reply app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt:756]: error - 
```
expected: RATIO_EXPRESSION
but was : OBJECTTYPE_NOT_SET
Expected :RATIO_EXPRESSION
Actual   :OBJECTTYPE_NOT_SET
```

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (MohamedMedhat1998) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: This fix makes the test case take forever when it is run using Espresso.  
@rt4914 any idea?

**Review by MohamedMedhat1998 — COMMENTED** [review]
Left a comment

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [reply app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: Okay, so I thought there's some problem with my laptop, I was trying running the test on espresso and it didn't stop for an hour

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt:872]: Why are we changing the test name here?

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: Do we need this?
I guess test can pass with `typeText` also.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
@prayutsu Please follow our wiki section - https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#tips-to-run-test-cases-in-both-espresso-and-robolectric
All the test can be fixed with it.

**Comment by rt4914** [comment]
Running late on PRs right now. So I will prioritise this once @anandwana001 and @MohamedMedhat1998 approves it.

Thanks.

**Comment by prayutsu** [comment]
> @prayutsu Please follow our wiki section - https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#tips-to-run-test-cases-in-both-espresso-and-robolectric
> All the test can be fixed with it.

I read the wiki tho

> @prayutsu Please follow our wiki section - https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#tips-to-run-test-cases-in-both-espresso-and-robolectric
> All the test can be fixed with it.

@anandwana001 The wiki section suggests using - 

1. `@Config` notation that we already have in this test file

2. `withEffectiveVisibility()`  but we don't have `visibility` issues here

3. `@LooperMode(LooperMode.Mode.PAUSED)` that we already have in this file

4. `doesNotExist` notation, but I think it has no link here

5. `testCoroutineDispatchers` , I tried fixing test cases with it but it didn't work, consider the below example  - 

```
  @Test
  fun testRatioInputView_withRatioHaving3Terms_clickSubmitButton_noErrorIsDisplayed() {
    ActivityScenario.launch(InputInteractionViewTestActivity::class.java)
    onView(withId(R.id.test_ratio_input_interaction_view))
      .perform(
        typeText(
          "1:2:3"
        )
      )
    testCoroutineDispatchers.runCurrent()
    closeSoftKeyboard()
    testCoroutineDispatchers.runCurrent()
    onView(withId(R.id.submit_button)).check(matches(isDisplayed())).perform(click())
    onView(withId(R.id.ratio_input_error)).check(matches(withText("")))
  }
```

**Comment by prayutsu** [comment]
@anandwana001 Am I still missing something?

**Comment by anandwana001** [comment]
> @anandwana001 Am I still missing something?

@prayutsu We need to run the test under launched intent using Closeable on this resource. 
Reference - https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html

**Comment by prayutsu** [comment]
@anandwana001 Tried implementing suggested changes in some of the test cases, but it's not working, can you PTAL?

**Review by anandwana001 — COMMENTED** [review]
> @anandwana001 Tried implementing suggested changes in some of the test cases, but it's not working, can you PTAL?

Always fix the other issues(ktlint, in this case) before any review. So, we can confirm what specific things are there causing the error. 

As you are asking for approach here, please use the draft PR process.
When you comment things like something is not working, please also comment, what all things you had tried, what you think could be a possible issue, What approaches you had tried already.

**Comment by prayutsu** [comment]
> > @anandwana001 Am I still missing something?
> 
> @prayutsu We need to run the test under launched intent using Closeable on this resource.
> Reference - https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html

@anandwana001 So, I tried this in a few test cases, and I don't have an exact idea why this isn't working.
And I have already tried fixing the test cases by using `appendText` so we can try that again if we aren't able to achieve this with the current approach

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: @anandwana001 Tried using closeable in this test case but it is failing

**Comment by prayutsu** [comment]
@anandwana001 
Screenshot from 2020-11-03 16-26-06

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: could you please comment here?

**Comment by anandwana001** [comment]
@prayutsu Couple of things, follow in this order:

## Formatting
 - We don't need of `activityScenario` variable. Example to follow everywhere. Do not import `ActivityScenario`, import `ActivityScenario.launch`. 
```diff
- val activityScenario = ActivityScenario.launch(....)
-    activityScenario.onActivity { activity ->
+ launch(...).use {
+      it.onActivity { activity ->
+      }
+ }
```

### Espresso
 - All test cases should pass in espresso. 
 - Currently, these 6 tests are failing in espresso
Screenshot 2020-11-09 at 09 12 59

Fixing one,

```diff
@Test
fun testRatioInputInteractionView_withInputtedText_hasCorrectPendingAnswer() {
-  val activityScenario = ActivityScenario.launch(
+ launch(
    InputInteractionViewTestActivity::class.java
  )
- activityScenario.use {
+ .use { it->
    onView(withId(R.id.test_ratio_input_interaction_view))
      .perform(
        typeText(
          "1:2:3"
        )
      )
- }
- activityScenario.onActivity { activity ->
-      val pendingAnswer = activity.ratioExpressionInputInteractionViewModel.getPendingAnswer()
-      assertThat(pendingAnswer.answer).isInstanceOf(InteractionObject::class.java)
-      assertThat(pendingAnswer.answer.objectTypeCase).isEqualTo(
-        InteractionObject.ObjectTypeCase.RATIO_EXPRESSION
-      )
-      assertThat(pendingAnswer.answer.ratioExpression.ratioComponentList)
-        .isEqualTo(listOf(1, 2, 3))
-    }

+    it.onActivity { activity ->
+      val pendingAnswer = activity.ratioExpressionInputInteractionViewModel.getPendingAnswer()
+      assertThat(pendingAnswer.answer).isInstanceOf(InteractionObject::class.java)
+      assertThat(pendingAnswer.answer.objectTypeCase).isEqualTo(
+        InteractionObject.ObjectTypeCase.RATIO_EXPRESSION
+      )
+      assertThat(pendingAnswer.answer.ratioExpression.ratioComponentList)
+        .isEqualTo(listOf(1, 2, 3))
+    }
  }
}
```

## Robolectric
- using delay from testCoroutineDispatchers might work.

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: @anandwana001 Tried implementing suggested changes, but it is still failing on robolectric

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: @anandwana001 I tried fixing test cases by `.runcurrent()` as well as `.advancedUntilIdle()` but none of them worked

**Review by prayutsu — COMMENTED** [review]
Adding a screenshot
Screenshot from 2020-11-12 13-14-53

**Comment by anandwana001** [comment]
unassigning myself, let me know when it is ready for review.

**Comment by prayutsu** [comment]
@anandwana001 PTAL

**Comment by anandwana001** [comment]
@prayutsu PR title is incorrect. 
Also update the screenshot and add both, robolectric and espresso

**Comment by prayutsu** [comment]
@anandwana001 Done

**Comment by anandwana001** [comment]
we are ignoring 7 test cases right!!
I will do a meeting with @rt4914 and will see what should be done ahead. By the time I think we need to keep this pr as `fix part of #1966` only, unless @prayutsu want to work ahead or leaving these test cases to be solved in other pr's.

**Comment by prayutsu** [comment]
@anandwana001 I think 3 or 4 test cases are being ignored because landscape mode is not implemented correctly yet, and the remaining 3-4 test cases can be fixed in other PRs as SLoP is about to finish and it would be good if I collect a few points and make it to Top 10.

**Comment by anandwana001** [comment]
> @anandwana001 I think 3 or 4 test cases are being ignored because landscape mode is not implemented correctly yet, and the remaining 3-4 test cases can be fixed in other PRs as SLoP is about to finish and it would be good if I collect a few points and make it to Top 10.

I will defer to @rt4914

**Comment by rt4914** [comment]
> > @anandwana001 I think 3 or 4 test cases are being ignored because landscape mode is not implemented correctly yet, and the remaining 3-4 test cases can be fixed in other PRs as SLoP is about to finish and it would be good if I collect a few points and make it to Top 10.
> 
> I will defer to @rt4914

Actually we can merge this PR as `part of ` but we cannot mark the original issue as closed and because of which we cannot provide points for this work.
I have filed 3 new issues recently I think they are easy and worth good points too.

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: Are we using this anymore?

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: Do we need intents?

**Review by anandwana001 — CHANGES_REQUESTED** [review]
@prayutsu PTAL

**Comment by anandwana001** [comment]
@prayutsu please prioritize this pr, as we are wrapping the robolectric stuff asap.

**Comment by prayutsu** [comment]
> @prayutsu please prioritize this pr, as we are wrapping the robolectric stuff asap.

Sure, will do that

**Comment by anandwana001** [comment]
@prayutsu update this branch with the latest develop branch.

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [reply app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: Done

File: app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt
Comment (prayutsu) [reply app/src/sharedTest/java/org/oppia/android/app/testing/InputInteractionViewTestActivityTest.kt]: Removed it

**Review by anandwana001 — APPROVED** [review]
LGTM, Thanks @prayutsu 

Defer to @rt4914 for further.

**Comment by anandwana001** [comment]
@prayutsu Before the merge, let's do one thing. All the tests which marked as ignore left them unchanged, so it can be done in another pr.

**Comment by prayutsu** [comment]
@anandwana001 PTAL

**Review by anandwana001 — CHANGES_REQUESTED** [review]
@prayutsu one last thing and after that, it is good to merge. 
Please revert the last commit, we want to describe why we are ignoring the test cases so it helps in future to solve.

**Comment by prayutsu** [comment]
> @prayutsu one last thing and after that, it is good to merge.
> Please revert the last commit, we want to describe why we are ignoring the test cases so it helps in future to solve.

Done

**Comment by anandwana001** [comment]
> > @prayutsu one last thing and after that, it is good to merge.
> > Please revert the last commit, we want to describe why we are ignoring the test cases so it helps in future to solve.
> 
> Done

It's looks incorrect!! What we are planning to have is only the edited ignore as they have comments, not thier specific modified tests.

**Comment by prayutsu** [comment]
@anandwana001 Can you please confirm once again if this looks correct now?

**Comment by anandwana001** [comment]
LGTM, Thanks

**Comment by rt4914** [comment]
@prayutsu You have done a force push just above this comment https://github.com/oppia/oppia-android/pull/2026#issuecomment-732060145 

You will need to close this PR and re-create. 

You can assign new PR to me and Mohamed that should be fine.

**Comment by rt4914** [comment]
Closing this because of force push.

**Pull request closed without merging**
